server/channels/api4/outgoing_oauth_connection.go

```diff
@@ -328,3 +333,60 @@ func deleteOutgoingOAuthConnection(c *Context, w http.ResponseWriter, r *http.Re
 
 	ReturnStatusOK(w)
 }
+
+// validateOutgoingOAuthConnectionCredentials validates the credentials of an outgoing oauth connection by requesting a token
+// with the provided connection configuration. If the credentials are valid, the request will return a 200 status code and
+// if the credentials are invalid, the request will return a 400 status code.
+func validateOutgoingOAuthConnectionCredentials(c *Context, w http.ResponseWriter, r *http.Request) {
+	auditRec := c.MakeAuditRecord("validateOutgoingOAuthConnectionCredentials", audit.Fail)
+	defer c.LogAuditRec(auditRec)
+	c.LogAudit("attempt")
+
+	if !checkOutgoingOAuthConnectionWritePermissions(c) {
+		w.WriteHeader(c.Err.StatusCode)
+		return
+	}
+
+	service, ok := ensureOutgoingOAuthConnectionInterface(c, whereOutgoingOAuthConnection)
+	if !ok {
+		w.WriteHeader(http.StatusNotImplemented)
+		return
+	}
+
+	// Allow checking connections sent in the body or by id if coming from an already existing
+	// connection url.
+	var inputConnection *model.OutgoingOAuthConnection
+	if c.Params != nil && c.Params.OutgoingOAuthConnectionID != "" {
+		var err *model.AppError
+		inputConnection, err = service.GetConnection(c.AppContext, c.Params.OutgoingOAuthConnectionID)
+		if err != nil {
+			c.Err = model.NewAppError(whereOutgoingOAuthConnection, "api.context.outgoing_oauth_connection.validate_connection_credentials.app_error", nil, err.Error(), http.StatusInternalServerError)
+			w.WriteHeader(c.Err.StatusCode)
+			return
+		}
+	} else {
+		if err := json.NewDecoder(r.Body).Decode(&inputConnection); err != nil {
+			c.Err = model.NewAppError(whereOutgoingOAuthConnection, "api.context.outgoing_oauth_connection.validate_connection_credentials.input_error", nil, err.Error(), http.StatusBadRequest)
+			w.WriteHeader(c.Err.StatusCode)
+			return
+		}
+	}
+
+	audit.AddEventParameterAuditable(auditRec, "outgoing_oauth_connection", inputConnection)
+
+	resultStatusCode := http.StatusOK
+
+	// Try to retrieve a token with the provided credentials
+	// do not store the token, just check if the credentials are valid and the request can be made
+	_, err := service.RetrieveTokenForConnection(c.AppContext, inputConnection)
+	if err != nil {
+		c.Logger.Error("Failed to retrieve token while validating outgoing oauth connection", logr.Err(err))
+		resultStatusCode = err.StatusCode
```

Comment:
Any reason for not returning here?

Reply:
Because of the audit. From my point of view the request itself was a success even if the token failed to retrieve, since we are just testing the connection, that's why I'm not failing early here.